# prunus-cp-genome-assembly
Project for Genomics Course 2021 (MSc Bioinformatics for Computational Genomics) helded by Prof. Aureliano Gomez Bombarely at Universit√† degli Studi di Milano.

## **Aim**
The project aims to investigate and compare the de novo assemblies of the chloroplast genome using both short-reads of [Illumina](https://www.illumina.com/) and long-reads coming from [PacBio](https://www.pacb.com/) and [Oxford Nanopore](https://nanoporetech.com/).

The chloroplast genome of Prunus avium was chosen for the assembly.

## **Description**
### 1. Retrieving raw data
The sequencing raw data were retrieved from the [SRA repository](https://www.ncbi.nlm.nih.gov/sra) in NCBI under the accession number:
* [SRR10362958](https://www.ncbi.nlm.nih.gov/sra/?term=SRR10362958): Illumina 
* [SRR4280451](https://www.ncbi.nlm.nih.gov/sra/?term=SRR4280451): PacBio SMRT
* [SRR7786091](https://www.ncbi.nlm.nih.gov/sra/?term=SRR7786091): Oxford Nanopore

The reference chloroplast genome of Prunus avium ([MK622380.1](https://www.ncbi.nlm.nih.gov/nuccore/MK622380.1)) and Prunus apetala ([NC_053693.1](https://www.ncbi.nlm.nih.gov/nuccore/NC_053693.1)) were taken from [NCBI Nucleotide Database](https://www.ncbi.nlm.nih.gov/nuccore).

The raw data were downloaded and converted into FASTQ files using `fastq-dump`.

The statistics of the fastq were obtained using [`fastq-stats`](https://expressionanalysis.github.io/ea-utils/).

### 2. Mapping
The mapping was performed both to the prunus avium and prunus apetala using:
* [`bowtie2`](https://github.com/BenLangmead/bowtie2) for short-reads 
* [`minimap2`](https://github.com/lh3/minimap2) and [`ngmlr`](https://github.com/philres/ngmlr) for long-reads

Illumina reads were pre-processed by using [`fastq-mcf`](https://expressionanalysis.github.io/ea-utils/) to remove the adapters before the mapping.

The reads mapping to the chloroplast genome were extracted using [`samtoos view`](http://www.htslib.org/doc/samtools-view.html).
The mapping stats were evaluated with [`samtools stats`](http://www.htslib.org/doc/samtools-stats.html).

### 3. Coverage evaluation
The mapped reads were sorted by position using [`samtool sort`](http://www.htslib.org/doc/samtools-sort.html).
[`bedtools genomecov`](https://bedtools.readthedocs.io/en/latest/content/tools/genomecov.html) with the option `-d` was used to obtain the depth at each genome position.

### 4. Assembly
The `BAM` files obtained from the mapping to Prunus avium were converted into FASTQ with [`bedtools bamtofast`](https://bedtools.readthedocs.io/en/latest/content/tools/bamtofastq.html).
In order to find the best assembly, several subsets of reads were generated by [`sektq sample`](https://github.com/lh3/seqtk).

The assemblies were obtained choosing:
* [`canu`](https://github.com/marbl/canu) for long-reads
* [`AbySS`](https://github.com/bcgsc/abyss) for short-reads comparing different kmer size

The statistics of each assembly were obtained with [`FastaSeqStats`](https://github.com/aubombarely/GenoToolBox/blob/master/SeqTools/FastaSeqStats).

### 5. Annotation of the longest contigs reconstructed
The longest contigs for each dataset were selected using [`FastaExtract`](https://github.com/aubombarely/GenoToolBox/blob/master/SeqTools/FastaExtract) and aligned to the reference using [BLASTN](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch).

The annotation was performed using [GeSeq](https://chlorobox.mpimp-golm.mpg.de/geseq.html).



